pkg/auth/providers/keycloakoidc/keycloak_client.go

```diff
@@ -0,0 +1,179 @@
+package keycloakoidc
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+)
+
+type searchResults struct {
+	Items []Account `json:"items"`
+}
+
+//Account defines properties an account in keycloak has
+type Account struct {
+	ID            int    `json:"id,omitempty"`
+	Email         string `json:email,omitempty`
+	EmailVerified string `json:emailVerified,omitempty`
+	Username      string `json:"username,omitempty"`
+	Enabled       bool   `json:"enabled,omitempty"`
+	Name          string `json:"firstName,omitempty"`
+	LastName      string `json:"lastName,omitempty"`
+	Type          string
+}
+
+//Group defines properties a group in keycloak has
+type Group struct {
+	ID        int     `json:"id,omitempty"`
+	Name      string  `json:"name,omitempty"`
+	Subgroups []Group `json:subGroups,omitempty`
+}
+
+//KClient implements a httpclient for keycloak
+type KClient struct {
+	httpClient *http.Client
+}
+
+func (k *KClient) searchPrincipals(searchTerm, principalType string, accessToken string, config *v32.OIDCConfig) ([]Account, error) {
+	var accounts []Account
+
+	if principalType == "" || principalType == UserType {
+		var userAccounts []Account
+		searchURL := fmt.Sprintf("%s/%ss?search=%s", k.getSearchURL(config.Issuer), UserType, searchTerm)
+		search := URLEncoded(searchURL)
+
+		b, err := k.getFromKeyCloak(accessToken, search)
+		if err != nil {
+			// no match on search returns an error. do not log
```

Comment:
Is there a specific error returned when there is no match? It is better to match against specific error code or string and decide to ignore, for other errors we need to bubble the error up.

Reply:
I think you and @cbron had similar questions - the error returned when nothing is found is a 401, with a message saying "nothing was found". We can match against that error code if we would like and then error error when other codes are retuned.

Reply:
Based on how the other providers are handling their get requests, it seems fairly standard to keep it is currently.

Reply:
We ned to handle the errors here so we don't eat them. It can be more intelligent and check error types but in general we need to return something.